### PR TITLE
feat(typography): add PRETTY/BALANCE textWrap modes with DP-based balanced wrapping

### PR DESCRIPTION
**Pull request by Vansh0204**
PR2
- Adds textWrap(PRETTY) and textWrap(BALANCE) to improve line-breaking aesthetics.
- Uses dynamic programming to minimize raggedness; last line has zero penalty.
- Integrates with existing text flow in 2D; works with all alignments, including JUSTIFIED .
- Lint/build pass locally.

Addresses and resloved #7712

### CODE REVIEW

**Comment by Vansh0204** [comment]
Hello @osteele @andrewn @zeke @reidab sir can u please review my PR.

**Comment by dhowe** [comment]
@Vansh0204 have you created tests for this feature?

**Comment by Vansh0204** [comment]
Thanks for reviewing, sir! I haven't added tests yet – I'll add comprehensive unit tests for the PRETTY/BALANCE textWrap modes covering the main functionality and edge cases. I'll push them shortly.

**Comment by dhowe** [comment]
a good place to start would be the alignment tests in `https://github.com/processing/p5.js/blob/dev-2.0/test/manual-test-examples/type/index.html`

image

eventually we will also need pixel-based tests like those here: https://github.com/processing/p5.js/blob/dev-2.0/test/unit/visual/cases/typography.js

**Comment by Vansh0204** [comment]
Hi @dhowe, sir! I've updated the PRs to include the necessary tests.
Screenshot 2025-12-01 at 12 05 08 AM

**Comment by dhowe** [comment]
Sorry, did you make this PR against the dev-2.0 branch ?

**Comment by Vansh0204** [comment]
Yes sir, the PR is currently targeting dev-2.0. Should I change the base branch to main instead, or would you prefer I rebase my changes on top of dev-2.0? Let me know which branch you'd like me to target, and I'll update accordingly.

File: src/core/constants.js [hunk 1]
```diff
@@ -1053,6 +1053,16 @@ export const CHAR = 'CHAR';
  * @final
  */
 export const WORD = 'WORD';
+/**
+ * @property {String} PRETTY
+ * @final
+ */
+export const PRETTY = 'pretty';
```
Comment (osteele) [inline src/core/constants.js]: p5.js uses uppercase e.g. `PRETTY = 'PRETTY'` by convention

**Comment by dhowe** [comment]
> Yes sir, the PR is currently targeting dev-2.0. Should I change the base branch to main instead, or would you prefer I rebase my changes on top of dev-2.0? Let me know which branch you'd like me to target, and I'll update accordingly.

I'm seeing main as the target for this pr:
image

**Comment by Vansh0204** [comment]
Yes @dhowe  sir, that's correct! The PR is targeting main. I can see it shows "base: main" at the top of the PR. Let me know if there's anything else you'd like me to address!

**Comment by dhowe** [comment]
both of these PR's need to target `dev-2.0` as we're not adding features to v1.x (which his still `main`) at this point

**Comment by Vansh0204** [comment]
@dhowe sir, I have to work on it.

**Comment by Vansh0204** [comment]
Hi @osteele and @dhowe, sir, I have raised new PRs can you please review them. Should I close these two PRs?

**Comment by dhowe** [comment]
> Hi @osteele and @dhowe, sir, I have raised new PRs can you please review them. Should I close these two PRs?

yes please

and thanks for your work!

**Comment by Vansh0204** [comment]
Thank you, @dhowe sir for your guidance.

**Comment by Vansh0204** [comment]
@dhowe Sir, one last request, can you please review my new PRs?
